Bug in compiling instance clauses

While running agda2hs to compile some code to Haskell the compilation process crashes with this log
```
...
Writing build/Data/Map/Internal/Mapping.hs
Writing build/Data/Map/Internal/Intersection.hs
agda2hs: Main.hs:437:7-68: Non-exhaustive patterns in p : ps

make: *** [Makefile:8: build] Error 1
```

The line in question is this 
```haskell
compileInstanceClause :: LocalDecls -> Clause -> TCM (Maybe (Hs.InstDecl ()))
compileInstanceClause ls c = do
  -- abuse compileClause:
  -- 1. drop any patterns before record projection to suppress the instance arg
  -- 2. use record proj. as function name
  -- 3. process remaing patterns as usual
  let (p : ps) = dropWhile (isNothing . isProjP) (namedClausePats c)
      c' = c {namedClausePats = ps}
      ProjP _ q = namedArg p

  -- We want the actual field name, not the instance-opened projection.
  (q, _, _) <- origProjection q
  ...
  ```
  I'm not sure how the instance arg clauses are being handled here. But the last file compiled in the log output and the one after it; both properly compile with agda itself. So errorful code is probably not the reason triggering this fault.

I did use the latest version of the agda2hs code in this case.

Would be happy to provide any more info to narrow down the cause.

--- a/Main.hs
+++ b/Main.hs
@@ -396,7 +396,7 @@ compile _ _ _ def = processPragma (defName def) >>= \ p ->
   where tag code = [(nameBindingSite $ qnameName $ defName def, code)]
 
 compileInstance :: Definition -> TCM [Hs.Decl ()]
-compileInstance def = do
+compileInstance def = setCurrentRange (nameBindingSite $ qnameName $ defName def) $ do
   ir <- compileInstRule [] (unEl . defType $ def)
   locals <- takeWhile (isAnonymousModuleName . qnameModule . fst)
           . dropWhile ((<= defName def) . fst)
@@ -434,19 +434,22 @@ compileInstanceClause ls c = do
   -- 1. drop any patterns before record projection to suppress the instance arg
   -- 2. use record proj. as function name
   -- 3. process remaing patterns as usual
-  let (p : ps) = dropWhile (isNothing . isProjP) (namedClausePats c)
-      c' = c {namedClausePats = ps}
-      ProjP _ q = namedArg p
+  case dropWhile (isNothing . isProjP) (namedClausePats c) of
+    []     -> genericDocError =<< fsep (pwords $ "Type class instances must be defined using copatterns and " ++
+                                                 "cannot be defined using helper functions or record expressions.")
+    p : ps -> do
+      let c' = c {namedClausePats = ps}
+          ProjP _ q = namedArg p
 
-  -- We want the actual field name, not the instance-opened projection.
-  (q, _, _) <- origProjection q
+      -- We want the actual field name, not the instance-opened projection.
+      (q, _, _) <- origProjection q
 
-  let uf = hsName (show (nameConcrete (qnameName q)))
-  (_ , x) <- compileClause ls uf c'
-  arg <- fieldArgInfo q
-  if visible arg
-    then return $ Just $ Hs.InsDecl () (Hs.FunBind () [x])
-    else return Nothing
+      let uf = hsName (show (nameConcrete (qnameName q)))
+      (_ , x) <- compileClause ls uf c'
+      arg <- fieldArgInfo q
+      if visible arg
+        then return $ Just $ Hs.InsDecl () (Hs.FunBind () [x])
+        else return Nothing
 
 fieldArgInfo :: QName -> TCM ArgInfo
 fieldArgInfo f = do

--- a/test/Fail/NonCopatternInstance.agda
+++ b/test/Fail/NonCopatternInstance.agda
@@ -1,0 +1,21 @@
+
+module Fail.NonCopatternInstance where
+
+record HasId (a : Set) : Set where
+  field id : a → a
+
+open HasId ⦃ ... ⦄
+
+{-# COMPILE AGDA2HS HasId class #-}
+
+data Unit : Set where
+  MkUnit : Unit
+
+{-# COMPILE AGDA2HS Unit #-}
+
+instance
+  UnitHasId : HasId Unit
+  UnitHasId = record { id = λ x → x }   -- NOT CORRECT
+  -- UnitHasId .id x = x                -- CORRECT
+
+{-# COMPILE AGDA2HS UnitHasId #-}

--- a/test/golden/NonCopatternInstance.err
+++ b/test/golden/NonCopatternInstance.err
@@ -1,0 +1,3 @@
+test/Fail/NonCopatternInstance.agda:17,3-12
+Type class instances must be defined using copatterns and cannot be
+defined using helper functions or record expressions.

